src/dsp_tools/commands/excel2json/new_lists.py

```diff
@@ -12,33 +12,105 @@
 from dsp_tools.models.exceptions import InputError
 
 
-def _fill_id_and_parent_id_columns(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
+def _construct_ids(excel_dfs: dict[str, dict[str, pd.DataFrame]]) -> dict[str, dict[str, pd.DataFrame]]:
+    filled = {
+        filename: {
+            sheet_name: _fill_one_df_auto_id_resolve_duplicates(df, _get_preferred_language(df.columns))
+            for sheet_name, df in sheets.items()
+        }
+        for filename, sheets in excel_dfs.items()
+    }
+    filled = _analyse_resolve_all_excel_duplicates(filled)
+    return _fill_all_excel_parent_id(filled)
+
+
+def _fill_all_excel_parent_id(excel_dfs: dict[str, dict[str, pd.DataFrame]]) -> dict[str, dict[str, pd.DataFrame]]:
+    return {
+        filename: {
+            sheet_name: _fill_one_df_parent_id(df, _get_preferred_language(df.columns))
+            for sheet_name, df in sheets.items()
+        }
+        for filename, sheets in excel_dfs.items()
+    }
+
+
+def _analyse_resolve_all_excel_duplicates(
+    excel_dfs: dict[str, dict[str, pd.DataFrame]],
+) -> dict[str, dict[str, pd.DataFrame]]:
+    ids = []
+    for sheets in excel_dfs.values():
+        for df in sheets.values():
+            ids.extend(df["id"].tolist())
+
+    id_df = pd.DataFrame({"id": ids})
+    if (duplicates := id_df["id"].duplicated(keep=False)).any():
+        id_set = set(id_df["id"][duplicates].to_list())
+        return _resolve_duplicates_in_all_excel(id_set, excel_dfs)
+    return excel_dfs
+
+
+def _resolve_duplicates_in_all_excel(
+    id_set: set[str], excel_dfs: dict[str, dict[str, pd.DataFrame]]
+) -> dict[str, dict[str, pd.DataFrame]]:
+    for filename, sheets in excel_dfs.items():
+        for sheet, df in sheets.items():
+            preferred_lang = _get_preferred_language(df.columns)
+            for i, row in df.iterrows():
+                if row["id"] in id_set:
+                    df.at[i, "id"] = _construct_non_duplicate_id_string(df.iloc[int(str(i))], preferred_lang)
+    return excel_dfs
+
+
+def _fill_one_df_auto_id_resolve_duplicates(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
```

Comment:
This function name is complicated, because the function makes 2 things at a time. Suggestion:
pack lines 65-68 into the function `_fill_one_df_auto_id()`. Then, you can call `_fill_one_df_auto_id()` and `_resolve_duplicate_custom_and_auto_id()` from one hierarchy above. So, this function here becomes obsolete.

Reply:
If you decide against that, I'd suggest to rephrase the function name so that it becomes an understandable sentence. Currently, it is a "heap of words", without syntactical coherence.